### PR TITLE
Added enum support in nimValueToPy

### PR DESCRIPTION
**Pull request by Clonkk**
Simple change to handle enum values as integer. 
Since `ord` it will be the integer value of the enum and not the string value

### CODE REVIEW

**Comment by Clonkk** [comment]
@Vindaar What do you think ?

**Comment by Vindaar** [comment]
In general I'm not sure if forcing enums to be converted by default (at least in such a way) is the best idea. Given that enums can have string values as well, the more "correct" approach would imo be to use its _value_. If the user gives the enum a string value it's likely there's a reason for that. 

And there may be different cases where one might want the ordinal value in one case and the string value in another. So forcing either isn't a clean solution to me. Enums are a bit special, because they sort of have 2 different interpretations of their values in Nim (or possibly even 3 if one includes parsing the enum field name from a string, but that's questionable, given that the same could be said even for numbers at that point).

But this is up to @yglukhov to decide anyway.

> The only issues with handling this way is that enum that are actually not consecutive ordinal may still have invalid values as the conversion to enum is a range check, so if your value is "hole" in your enum the conversion will still compile.

This I don't quite understand. The conversion in your PR doesn't have anything to do with how the value ends up in the `nimValueToPy` proc in the first place. Your code that produces the value needs to make sure it doesn't produce invalid enum values. If you just use `MyEnum(<invalid enum int>)` that's problematic, but not nimpy's task to raise an error here.

**Comment by Clonkk** [comment]
> I'm not sure if forcing enums to be converted by default (at least in such a way) is the best idea. Given that enums can have string values as well

This bit in the [manual](https://nim-lang.org/docs/manual.html#types-enumeration-types) :
> The compiler supports the built-in stringify operator $ for enumerations. The stringify's result can be controlled by explicitly giving the string values to use.

If I understand it correctly, an enum doesn't have a string value per se but have a stringified value that you can specify.
The string "value" is only used for stringify (`$`) but it doesn't change the enum nature : it is still an Ordinal (and can be used with `SomeOrdinal` ).
For example : 
```nim
type
  MyEnum = enum
    m1 = "a1"
    m2 = "a2"
    m3 = "a3"

## This works
echo MyEnum(0)
echo MyEnum(1)
echo MyEnum(2)

## This does not compile
echo MyEnum("a1")
```
Which I think highlights that an enum shouldn't be "attached" to its stringify value (you can't create an enum value from its string value).

What would be doable is to interpret an enum as a tuple (string, int) like this : `nimTupleToPy(($v, ord(v))`.
But I think it would be confusing in the case where your enum don't have a string value.
It also makes the enum a bit more complicated to use in Python as you suddenly have a tuple of 2 value 
```python
{"param4": ("a2": 1)}
```
> The conversion in your PR doesn't have anything to do with how the value ends up in the nimValueToPy proc in the first place. Your code that produces the value needs to make sure it doesn't produce invalid enum values. If you just use MyEnum(<invalid enum int>) that's problematic, but not nimpy's task to raise an error here.

Yeah, it was out of topic.

**Comment by Vindaar** [comment]
> If I understand it correctly, an enum doesn't have a string value per se but have a stringified value that you can specify.
> The string "value" is only used for stringify (`$`) but it doesn't change the enum nature : it is still an Ordinal (and can be used with `SomeOrdinal` ).
> For example :
> 
> ```nim
> type
>   MyEnum = enum
>     m1 = "a1"
>     m2 = "a2"
>     m3 = "a3"
> 
> ## This works
> echo MyEnum(0)
> echo MyEnum(1)
> echo MyEnum(2)
> 
> ## This does not compile
> echo MyEnum("a1")
> ```
> 
> Which I think highlights that an enum shouldn't be "attached" to its stringify value (you can't create an enum value from its string value).

Not quite. That's precisely why `strutils.parseEnum` exists:
https://nim-lang.github.io/Nim/strutils.html#parseEnum%2Cstring

It is simply extremely useful to be able to define a specific set of strings, without having to resort to defining string constants or, even worse, hardcoding the strings in the code. Especially, because using the enum values throughout the code and accessing the string values when desired allows for compile time errors when using an invalid field.

The issue I see with this approach is that now enums are fixed as their ordinal values. Same as there isn't a way to hand an enum to Python without going through trouble converting it beforehand on the Nim side before this PR, after merging this there is still no way to hand the string value of enums to Python. But worse, because now one does not even get a compile time error.

That's why I think an approach that allows the _user to choose_ is the best, because there are good use cases for wanting the ordinal value as well as the string value.

> 
> Another doable thing is to interpret an enum as a tuple (string, int) like this : `nimTupleToPy(($v, ord(v))`.
> But I think it would be confusing in the case where you haven't defined a string value for your enum.
> It also makes the enum a bit more complicated to use in Python as you suddenly have a tuple of 2 value
> 
> ```python
> {"param4": ("a2": 1)}
> ```
> 
> In my opinion, treating the enum as an integer is both simpler to use and understand but maybe that's because I'm used to C.

I agree, that this would be confusing in the general case. 

> Is string enum a thing in Python ?

All I know about enums in Python is the 2 minutes I googled earlier today.

**Comment by Clonkk** [comment]
So, I've done some test and I think the best solution is a generic proc `nimpyEnumConvert` that can be overloaded. 

That way we can have a default ìnt` conversion and let the user overload the enum (or all) he wishes. 

I'm still trying to test the limit of the overloading syntax see if I can do a more generic version.

**Comment by Clonkk** [comment]
@yglukhov @Vindaar Any thoughts ?

**Comment by Vindaar** [comment]
Sorry, am super busy the last few days and wanted to leave yglukhov to review.

Just a couple of things to note (no ill intentions):
1. please remove the style changes introduced by (I assume) running `nimpretty`. That makes reviewing the changes really hard and also those things are up to the repo owner to decide.
2. What's that first overload that's just equal to `delete`?
3. without having thought deeply about this: overloading the proc works correctly? If so that seems to me to go the direction I had in mind.

**Comment by Clonkk** [comment]
> 1.

Fair. I forgot I turned it on automatically.

> 2.

Simplest way I could find to allow an overload where the parameters are identical and the return type only is different. Since there is no return type inference in Nim, overload that  only change return value would leads to ambiguous call.

>3.
I added a unit test that checks non-overloaded enum get passed as their ordinal value and overloaded enum get passed as their string value.

**Comment by Vindaar** [comment]
> Simplest way I could find to allow an overload where the parameters are identical and the return type only is different. Since there is no return type inference in Nim, overload that only change return value would leads to ambiguous call.

I don't get it. The signature is different (2 args instead of 1). And I still don't understand what a body containing `delete` is supposed to mean.

> I added a unit test that checks non-overloaded enum get passed as their ordinal value and overloaded enum get passed as their string value.

Cool!

File: nimpy.nim
Comment (yglukhov) [inline nimpy.nim:848]: This doesn't seem to be needed. `nimpyEnumConvert` is never called with 2 arguments.

File: tests/py_enum_overload.py
Comment (yglukhov) [inline tests/py_enum_overload.py:1]: It is beneficial to add the tests to existing `pyfromnim` tests instead of creating new files. That's because pyfromnim is tested in both python-originated environment (when nim is python module) and python-imported environment (when nim is executable).

**Review by yglukhov — CHANGES_REQUESTED** [review]

**Comment by Clonkk** [comment]
I'll try to rework this a bit because I realized that my proposed solution only works if the overloaded function is defined in the same file as callMethod so you can't put it in a module.
The lack of return value overloading is not making this easy to solve.

I keep in mind the change requested and recommandation made.

**Comment by yglukhov** [comment]
> only works if the overloaded function is defined in the same file

Have you tried using `mixin` statement?

**Comment by Clonkk** [comment]
> Have you tried using mixin statement?

Didn't know about that. Yup it works.

**Comment by yglukhov** [comment]
Thanks!

**Pull request merged**
